### ISSUE TITLE
Issues with DRF3 and HyperlinkedIndentityField

### ISSUE THREAD

**Issue by cancan101** [alanjds/drf-nested-routers#38]
See https://github.com/tomchristie/django-rest-framework/issues/1773#issuecomment-63154330

**Comment by alanjds** [comment]
Sorry for the delayed response. Can you please contribute a PR ?

**Comment by hvdklauw** [comment]
This is what I've implemented for this, just a small note that it only allows for the first layer of nesting and with simple getattr on the object to get the value for the resolving.

To get deeper nesting or some complex relationships the parent_* properties should be tuple/arrays and use django's double underscore method to transfer relationships.

```python
class NestedHyperlinkedIdentityField(HyperlinkedIdentityField):
    """HyperlinkedIdentifyField for nested relations."""

    parent_lookup_field = None
    parent_lookup_url_kwarg = None

    def __init__(self, *args, **kwargs):
        """Constructor."""
        self.parent_lookup_field = kwargs.pop('parent_lookup_field', None)
        self.parent_lookup_url_kwarg = kwargs.pop('parent_lookup_url_kwarg', None) or self.parent_lookup_field
        super(NestedHyperlinkedIdentityField, self).__init__(*args, **kwargs)

    def get_url(self, obj, view_name, request, format):
        """Get the url for the field."""
        # Unsaved objects will not yet have a valid URL.
        if hasattr(obj, 'pk') and obj.pk is None:
            return None

        lookup_value = getattr(obj, self.lookup_field)
        kwargs = {self.lookup_url_kwarg: lookup_value}
        if self.parent_lookup_field:
            kwargs[self.parent_lookup_url_kwarg] = getattr(obj, self.parent_lookup_field)
        return self.reverse(view_name, kwargs=kwargs, request=request, format=format)
```

Example usage for the nameservers example in the readme:
On the NameServer serializer:
```python

url = NestedHyperlinkedIdentityField(
    view_name='domain-nameservers-detail',
    lookup_field='pk',
    parent_lookup_field='domain_id',
    parent_lookup_kwarg='domain_pk'
)
```

**Comment by alanjds** [comment]
Looks nice, and thanks for sharing. Can you contribute a test too?

If not, I will merge this but not issue a release until some people confirm that it works and will not break existing apps.

**Comment by hvdklauw** [comment]
Yes, I have some time tomorrow, I'll see if I can turn it into a nice PR.

**Comment by alanjds** [comment]
Thanks a lot.

**Comment by hvdklauw** [comment]
Sorry, bit busy at work, did have a look, was hoping I could copy a test for the NestedHyperlinkedRelationField, but couldn't find it. So it will take a bit longer.

**Comment by alanjds** [comment]
No problem. Everyone needs to work.

There is tests, but not for Fields. You can grab some template there...

**Comment by mbaragiola** [comment]
@hvdklauw your solution works for me. Thanks for sharing.

**Comment by mbaragiola** [comment]
Hello, I've tried to extend @hvdklauw 's solution to a 3 level deep nested router. Here's the code:

    class NestedHyperlinkedIdentityField(HyperlinkedIdentityField):
        """
        """
        parent_parent_lookup_field = None
        parent_parent_lookup_url_kwarg = None
        parent_lookup_field = None
        parent_lookup_url_kwarg = None

        def __init__(self, *args, **kwargs):
            """Constructor."""
            self.parent_lookup_field = kwargs.pop('parent_lookup_field', None)
            self.parent_lookup_url_kwarg = (
                kwargs.pop('parent_lookup_url_kwarg', None) or (
                    self.parent_lookup_field
                )
            )
            self.parent_parent_lookup_field = kwargs.pop(
                'parent_parent_lookup_field',
                None
            )
            self.parent_parent_lookup_url_kwarg = (
                kwargs.pop('parent_parent_lookup_url_kwarg', None) or (
                    self.parent_parent_lookup_field
                )
            )

            super(NestedHyperlinkedIdentityField, self).__init__(*args, **kwargs)

    def get_url(self, obj, view_name, request, format):
            """Get the url for the field."""
            # Unsaved objects will not yet have a valid URL.
            if hasattr(obj, 'pk') and obj.pk is None:
                return None

            lookup_value = getattr(obj, self.lookup_field)
            kwargs = {self.lookup_url_kwarg: lookup_value}
            if self.parent_lookup_field:
                kwargs[self.parent_lookup_url_kwarg] = (
                    getattr(obj, self.parent_lookup_field)
                )
            if self.parent_parent_lookup_field:
                kwargs[self.parent_parent_lookup_url_kwarg] = (
                    getattr(obj, self.parent_parent_lookup_field)
                )

            return self.reverse(
                view_name,
                kwargs=kwargs,
                request=request,
                format=format
            )

The 2 level deep continues to work with this solution, but a 3 level deep one throws the lookup_field error. Django shows me that the router is generating URLs and viewnames as it should.

Extending the example in the README:
    class MailDropSerializer(HyperlinkedModelSerializer):
        recipients = NestedHyperlinkedIdentityField(
                view_name='recipients-list',
                parent_parent_lookup_url_kwarg='client_pk',
                parent_parent_lookup_field='client_id',
                parent_lookup_url_kwarg='maildrop_pk',
                parent_lookup_field='pk'
        )

Any ideas why isn't this working? Thanks in advance.

**Comment by hvdklauw** [comment]
The reason is that the obj does have the parent_lookup_field attribute, but
not the parent_parent_lookup_field, you need to get that from the
obj.parent, but for that you need to know how to traverse the relationship.

Also you should mind the extra query this will do unless you prefetch

**Comment by alanjds** [comment]
@mbaragiola I just updated your comment to add colors, ok?

Hey, can you please issue a PR, if this fixes someone problems without creating more problems?

**Comment by mbaragiola** [comment]
@alanjds you mean the 2 level deep NestedHyperlinkIdentityField?

**Comment by alanjds** [comment]
Yes. The multilevel field.

**Comment by alanjds** [comment]
Just released v0.90.0 with #91, and it could had fixed this one. Please take a look.
